### PR TITLE
SDK-1549 Reconnect restarts subscribe loop regardless of the state

### PR DESCRIPTION
**Pull request by kleewho**
fix: Reconnect always restarts the subscription connection

### CODE REVIEW

**Review by bartk — DISMISSED** [review]
LGTM but there is one question.

File: src/main/java/com/pubnub/api/builder/dto/PubSubOperation.java [hunk 1]
```diff
@@ -28,4 +29,12 @@ private ConnectedStatusAnnouncedOperation() {
 
         }
     }
+
+    @Data
+    class ReconnectOperation implements PubSubOperation {
+        private ReconnectOperation() {
```
[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/java/com/pubnub/api/builder/dto/PubSubOperation.java:35]: I'll check it and address it in later change if necessary

**Comment by kleewho** [comment]
@client-engineering-bot help

**Comment by kleewho** [comment]
@client-engineering-bot release

File: src/test/java/com/pubnub/api/managers/FastSubscriptionManagerTest.java [hunk 2]
```diff
@@ -65,6 +67,30 @@ public void performsLongPollingAfterTimeout() throws IllegalAccessException {
...
+        final SubscriptionManager subscriptionManager = subscriptionManagerUnderTest(retrofitManagerMock);
+
+        final SubscribeOperation subscribeOperation = SubscribeOperation.builder()
+                .channels(singletonList("ch1"))
+                .channelGroups(emptyList())
+                .build();
+
+        subscriptionManager.adaptSubscribeBuilder(subscribeOperation);
+
+        subscriptionManager.reconnect();
```
Comment (budgetpreneur) [inline src/test/java/com/pubnub/api/managers/FastSubscriptionManagerTest.java:90]: To my knowledge, reconnect happens only when it is disconnected. Without disconnect, is it logical to reconnect?

Comment (kleewho) [reply src/test/java/com/pubnub/api/managers/FastSubscriptionManagerTest.java:90]: Yes, it's logical. User can call this regardless of the state

**Review by bartk — APPROVED** [review]

File: src/test/java/com/pubnub/api/managers/FastSubscriptionManagerTest.java (hunk 2 above)
Comment (budgetpreneur) [reply src/test/java/com/pubnub/api/managers/FastSubscriptionManagerTest.java:90]: what happens when reconnect is called multiple times?

**Review by kazydek — APPROVED** [review]

**Review by techwritermat — APPROVED** [review]

File: src/test/java/com/pubnub/api/managers/FastSubscriptionManagerTest.java (hunk 2 above)
Comment (kleewho) [reply src/test/java/com/pubnub/api/managers/FastSubscriptionManagerTest.java:90]: We will stop any currently ongoing call, and restart the loop

Comment (budgetpreneur) [reply src/test/java/com/pubnub/api/managers/FastSubscriptionManagerTest.java:90]: ok. thanks for clarifying.

Comment (kleewho) [reply src/test/java/com/pubnub/api/managers/FastSubscriptionManagerTest.java:90]: No problem!

**Pull request merged**